polkadot/xcm/xcm-builder/src/matches_location.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Various implementations and utilities for matching and filtering `MultiLocation` and
+//! `InteriorMultiLocation` types.
+
+use frame_support::traits::{Contains, Everything, Get};
+use xcm::latest::{InteriorMultiLocation, MultiLocation};
+
+/// Trait for matching location of type `T`.
+pub trait MatchesLocation<T> {
+	fn matches(&self, location: &T) -> bool;
+}
+
+/// A [`MatchesLocation`] implementation that matches every value.
+impl<T> MatchesLocation<T> for Everything {
+	fn matches(&self, _: &T) -> bool {
+		true
+	}
+}
+
+/// Adapter for using `trait Contains` with `trait MatchesLocation` were we can provide a tuple of
+/// `Contains` implementations which are used for matching location of type `T`.
+pub struct MatchesLocationAdapter<T, Filter> {
+	_marker: sp_std::marker::PhantomData<(T, Filter)>,
+}
+impl<T, Filter> MatchesLocationAdapter<T, Filter>
+where
+	Filter: Contains<T>,
+{
+	pub fn new() -> Self {
+		Self { _marker: sp_std::marker::PhantomData }
+	}
+}
+impl<T, Filter> MatchesLocation<T> for MatchesLocationAdapter<T, Filter>
+where
+	Filter: Contains<T>,
+{
+	fn matches(&self, location: &T) -> bool {
+		Filter::contains(location)
+	}
+}
```

Comment:
Do we need `MatchesLocation` ? It looks like it has the same definition as `Contains`, only with different naming and it's doing exactly what `Contains` does. Can we just use `Contains` directly ?

Reply:
Yes, I know exactly what you mean, but the difference with `Contains` is `&self`.
The main idea was to use tuples of `Contains` for filtering, but I didnt find a way how to pass `Contains` impl to the `Vec<NetworkExportTableItem>` [here](https://github.com/paritytech/polkadot-sdk/pull/1848/files#diff-47dce7275c4bcc7e3b9a8486c2635fd1b383ea740774912705857aa5a79a6fb2R155-R165).
So this trait `MatchesLocation` and `MatchesLocationAdapter`, which wraps tuples of  `Contains` impls, can be passed as instance to the `Vec`.
This trait is also reused for other scenarios [here](https://github.com/paritytech/polkadot-sdk/pull/1215).

Reply:
Oh, I see. Sorry, I missed this. Thinking.

Reply:
One option to make this logic more generic would be something like:

```
pub struct PhantomInstance<T>(sp_std::marker::PhantomData<T>);

impl<T> PhantomInstance<T> {
	pub fn new() -> Self {
		Self(Default::default())
	}
}

pub trait SelfContains<T> {
	fn contains(&self, t: &T) -> bool;
}

impl<T, C> SelfContains<T> for PhantomInstance<C>
where
	C: Contains<T>,
{
	fn contains(&self, t: &T) -> bool {
		C::contains(t)
	}
}
```

Which would result in changes like the following:
```
type AllowAll = InteriorLocationMatcher<Everything>; 
-> 
type AllowAll = PhantomInstance::<Everything>::new()


MatchesLocationAdapter::<
					Loc,
					(AllowOnly2000And3000And4000, Between5000And5003)
				>::new()
->
PhantomInstance::<(AllowOnly2000And3000And4000, Between5000And5003)>::new()
```

Not sure about the naming, but personally I like this a bit better, because it's more generic.

However, I'm still thinking if there isn't a better solution.

Reply:
Actually another option would be to define `remote_location_filter` as a `Option<Vec<InteriorMultiLocation>>` or `Option<HashSet<InteriorMultiLocation>>` instead of a filter. Would this be scalable enough ? Are there any use cases that would require a lot more flexibility ?

Reply:
> Actually another option would be to define `remote_location_filter` as a `Option<Vec<InteriorMultiLocation>>` or `Option<HashSet<InteriorMultiLocation>>` instead of a filter. Would this be scalable enough ? Are there any use cases that would require a lot more flexibility ?

Actually, you're right, let's not complicate this, I will change it tomorrow to the `Option<Vec<`, that should simplify and cover all cases, thanks :)

Reply:
@serban300 fixed and removed all that stuff [here](https://github.com/paritytech/polkadot-sdk/pull/1848/commits/5fdd7870031f68e3e1717a5139607afa52603273)